service/model/environment.go

```diff
@@ -0,0 +1,133 @@
+package model
+
+import (
+	"context"
+	"errors"
+	"net/url"
+	"time"
+
+	urlx "release-manager/pkg/url"
+
+	"github.com/google/uuid"
+)
+
+var (
+	errEnvironmentURLMustBeAbsolute = errors.New("provided url must be valid absolute url")
+	errEnvironmentNameRequired      = errors.New("environment name is required")
+)
+
+type EnvironmentRepository interface {
+	Create(ctx context.Context, env Environment) error
+	Read(ctx context.Context, projectID, envID uuid.UUID) (Environment, error)
+	ReadAllForProject(ctx context.Context, projectID uuid.UUID) ([]Environment, error)
+	Delete(ctx context.Context, projectID, envID uuid.UUID) error
+	Update(ctx context.Context, env Environment) error
+}
+
+type Environment struct {
+	ID        uuid.UUID
+	ProjectID uuid.UUID
+	Name      string
+	URL       url.URL
+	CreatedAt time.Time
+	UpdatedAt time.Time
+
+	// TODO add release id once releases are implemented
+}
+
+type EnvironmentCreation struct {
+	ProjectID uuid.UUID
+	Name      string
+	RawURL    string
+}
+
+type EnvironmentUpdate struct {
+	ID        uuid.UUID
+	ProjectID uuid.UUID
+	Name      *string
+	RawURL    *string
+}
+
+func NewEnvironment(projectID uuid.UUID, name, rawURL string) (Environment, error) {
```

Comment:
Why do I need to provide this URL? It should be accesible on the base url + id, shouldn't it?

Reply:
Renamed to `ServiceURL` to improve clarity.